### PR TITLE
[Bugfix][2822] Sound off when minimized

### PR DESCRIPTION
**Pull request by tau3**
Hello!

This a _partial_ fix for #2822 (MacOS/SDL2/**SDL_Mixer**)

What was done:

1. SDL2 window event handling logic was moved into a separate function (`OnSdl2WindowEvent`) beause `LocalEvent::HandleEvents` is too big and complex;
2. Changed detected events types from window hidden/shown to focus lost/acuired;
3. Unfortunately, SDL_Mixer doesn't pause played music _immediately_. I didn't consider halting and restarting music on focus change because it would restart a song each time - probably not good. Instead, here is a workaround - store current volume, set to zero and restore when focus is aquired

### CODE REVIEW

File: src/engine/localevent.h [hunk 1]
```diff
@@ -343,6 +344,7 @@ class LocalEvent
     bool _isHiddenWindow;
     bool _isMusicPaused;
     bool _isSoundPaused;
+    u16 _volume;
```
Comment (ihhub) [inline src/engine/localevent.h:347]: Please use `uint16_t` instead of **u16** and fix the build for SDL1.

File: src/engine/localevent.cpp [hunk 2]
```diff
@@ -52,6 +52,7 @@ LocalEvent::LocalEvent()
     , _isHiddenWindow( false )
     , _isMusicPaused( false )
     , _isSoundPaused( false )
+    , _volume( 0 )
```
Comment (ihhub) [inline src/engine/localevent.cpp:55]: Could we please be more explicit and define where this volume belongs to: sound or music?

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @tau3 , could you please take a look at given comment and also make changes for SDL 1 code as well? To remove code duplication we could introduce 2 private methods for pausing sounds and resuming them.

**Comment by tau3** [comment]
Hi @ihhub!

Thank you for reviews, as usual.

SDL1 build fails on my Big Sur - i wouldn't be able to check if changes work fine, that's why SDL1-related logic is left untouched.
Btw, I can't see why SDL1 build is broken after my recent commit. Do some access rights required?

**Comment by ihhub** [comment]
Hi @tau3 , to check what's wrong with the build please click on Details right from it. There you can check logs.
It fails due to unused variable.

File: src/engine/localevent.h [hunk 3]
```diff
@@ -269,6 +269,7 @@ class LocalEvent
     void OpenController();
     void CloseController();
     void OpenTouchpad();
+    void OnSdl2WindowEvent( const SDL_Event & event );
```
Comment (ihhub) [inline src/engine/localevent.h:272]: Please move this method into private section.

File: src/engine/localevent.cpp [hunk 4]
```diff
@@ -1130,6 +1111,33 @@ bool LocalEvent::HandleEvents( bool delay, bool allowExit )
 }
 
 #if SDL_VERSION_ATLEAST( 2, 0, 0 )
+void LocalEvent::OnSdl2WindowEvent( const SDL_Event & event )
+{
+    if ( Mixer::isValid() ) {
```
Comment (ihhub) [inline src/engine/localevent.cpp:1116]: Please modify the code to have fast return:
```
if ( !Mixer::isValid() ) {
    return;
}
```
so the rest of code will not be shifted much.

File: src/engine/localevent.cpp [hunk 5]
```diff
@@ -1130,6 +1111,33 @@ bool LocalEvent::HandleEvents( bool delay, bool allowExit )
...
+        if ( event.window.event == SDL_WINDOWEVENT_FOCUS_LOST ) {
+            _isHiddenWindow = true;
+            _isMusicPaused = Music::isPaused();
+            _isSoundPaused = Mixer::isPaused( -1 );
+            Mixer::Pause();
+            Music::Pause();
+            _musicVolume = Music::Volume( 0 );
+            loop_delay = 100;
```
Comment (ihhub) [inline src/engine/localevent.cpp:1124]: This section of the code is duplicated at lines 1002-1007. Could we please create another private method, let's call it `void StopSounds()` where we put all the logic there?

File: src/engine/localevent.cpp [hunk 6]
```diff
@@ -1130,6 +1111,33 @@ bool LocalEvent::HandleEvents( bool delay, bool allowExit )
...
+            if ( _isHiddenWindow ) {
+                if ( !_isMusicPaused ) {
+                    Music::Volume( _musicVolume );
+                    Music::Resume();
+                }
+                if ( !_isSoundPaused )
+                    Mixer::Resume();
+                _isHiddenWindow = false;
+            }
+            loop_delay = 1;
```
Comment (ihhub) [inline src/engine/localevent.cpp:1136]: This section of the code is duplicated at lines 1010-1017. Could we please create another private method, let's call it `void ResumeSounds()` where we put all the logic there?

I'm aware of music volume changes so could you please check if it affects SDL1?

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @tau3 , I left few comments about code structure. Could you please take a look?

File: src/engine/localevent.cpp (hunk 6 above)
Comment (tau3) [reply src/engine/localevent.cpp]: music keeps playing with SDL1 ¯\_(ツ)_/¯

Comment (tau3) [reply src/engine/localevent.cpp]: also `SDL_APPACTIVE` is only registered for `cmd + tab` switching. `cmd + h` doesn't work at all, and on mouse window switching or restoring from dock there are only a `SDL_JoyAxisEvent` events, so a workaround will be required in the future

File: src/engine/localevent.cpp [hunk 7]
```diff
@@ -1129,7 +1098,51 @@ bool LocalEvent::HandleEvents( bool delay, bool allowExit )
...
+{
+    _isHiddenWindow = true;
+    _isMusicPaused = Music::isPaused();
+    _isSoundPaused = Mixer::isPaused( -1 );
+    Mixer::Pause();
+    Music::Pause();
+
 #if SDL_VERSION_ATLEAST( 2, 0, 0 )
+    _musicVolume = Music::Volume( 0 );
+#endif
```
Comment (ihhub) [inline src/engine/localevent.cpp:1111]: What if we execute this code on SDL 1? If there are no changes then we don't need if-def blocks around `_musicVolume` variable.

Comment (tau3) [reply src/engine/localevent.cpp:1111]: i tried yesterday... this trick doesn't work on SDL1

Comment (ihhub) [reply src/engine/localevent.cpp:1111]: In this case can we just remove if-def if it does nothing?

**Comment by ihhub** [comment]
@tau3 , thank you very much for this fix!

**Pull request merged**